### PR TITLE
Added coalesce function for optionals

### PR DESCRIPTION
**Pull request by JadenGeller**

### CODE REVIEW

**Comment by CodaFi** [comment]
Thus begins the nitpicking session.

File: Swiftz/OptionalExt.swift [hunk 1]
```diff
@@ -37,3 +37,25 @@ public func getOrElse<T>(m : Optional<T>) -> T -> T {
...
+public func coalesce<T>(approve: T -> Bool) -> (@autoclosure () -> T? ...) -> T? {
+	return { all in
+		for f: () -> T? in all {
+			if let x = f() {
+				if approve(x) { return x }
+			}
+		}
+		return nil
+	}
+}
```
Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: Newline here.

File: Swiftz/OptionalExt.swift [hunk 2]
```diff
@@ -37,3 +37,25 @@ public func getOrElse<T>(m : Optional<T>) -> T -> T {
 		}
 	}
 }
+
+/// Given one or more Optional values, returns the first Optional value that is not nil
+/// when evaulated from left to right
+public func coalesce<T>(all: @autoclosure () -> T? ...) -> T? {
```
Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: As a matter of style, we separate name, colon, and identifier with spaces

```
public func coalesce<T>(all : @autoclosure () -> T? ...) -> T? {
```

File: Swiftz/OptionalExt.swift [hunk 3]
```diff
@@ -37,3 +37,25 @@ public func getOrElse<T>(m : Optional<T>) -> T -> T {
...
+	for f: () -> T? in all {
```
Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: Same here.

File: Swiftz/OptionalExt.swift [hunk 4]
```diff
@@ -37,3 +37,25 @@ public func getOrElse<T>(m : Optional<T>) -> T -> T {
...
+		if let x = f() { return x }
+	}
+	return nil
+}
+
+/// Given one or more Optional values, returns the first Optional value that is not nil
+/// and satisfies the approve function when evaulated from left to right
+public func coalesce<T>(approve: T -> Bool) -> (@autoclosure () -> T? ...) -> T? {
```
Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: And here.

File: Swiftz/OptionalExt.swift [hunk 5]
```diff
@@ -37,3 +37,25 @@ public func getOrElse<T>(m : Optional<T>) -> T -> T {
...
+	return { all in
+		for f: () -> T? in all {
```
Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: And here.

File: Swiftz/OptionalExt.swift (hunk 3 above)
Comment (JadenGeller) [inline Swiftz/OptionalExt.swift]: There's a bug in the current version of swift that prevents this inference.

File: Swiftz/OptionalExt.swift (hunk 1 above)
Comment (JadenGeller) [inline Swiftz/OptionalExt.swift]: I'm having a little bit of trouble understanding where you are wanting the newlines. From what I can tell it is consistent with the style of the rest of the codebase.

Comment (CodaFi) [inline Swiftz/OptionalExt.swift]: Oh, silly me. The `and here` comments are all about the above comment about separating identifier, colon, and type with spaces.  The only newline I want is at the end of this file.

**Comment by CodaFi** [comment]
This endeth the nitpicking session.

I know I've already said it once, but thank you for your contribution and taking the time to draw up this pull request.

:boat:

**Pull request merged**